src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs

```diff
@@ -1722,11 +1729,51 @@ var outboundNavigations
 
             foreach (var navigation in outboundNavigations)
             {
-                var addedIncludeTreeNode = includeTreeNode.AddNavigation(navigation);
-                PopulateEagerLoadedNavigations(addedIncludeTreeNode);
+                includeTreeNode.AddNavigation(navigation);
             }
         }
 
+        private void VerifyNoAutoIncludeCycles(
+            IEntityType entityType, HashSet<IEntityType> visitedEntityTypes, INavigationBase previousNavigation)
+        {
+            if (_nonCyclicAutoIncludeEntityTypes.Contains(entityType))
+            {
+                return;
+            }
+
+            if (!visitedEntityTypes.Add(entityType))
```

Comment:
```suggestion
            visitedEntityTypes.Add(entityType, previousNavigation);
```

Change `visitedEntityTypes` to dictionary and now you can reconstruct the cycle (more or less)